### DISCUSSION TITLE
短路如何处理

### DISCUSSION THREAD

**Discussion by hodanna7** [arcsysu/SYsU-lang#49] [Q&A]
我的想法：
对于if( exp1 && exp2){}，先对exp1进行br：CreateCondBr(exp1, rhsBB, nextBB)，其中rhsBB是处理exp2的块，nextBB是if语句后的块，然后在rhsBB中对exp2进行br：CreateCondBr(exp2, thenBB, nextBB)，其中thenBB是if条件成立后执行的块。
但是感觉这个想法实现有点复杂，不确定是否正确，而且如果涉及到else更加复杂，基本块比较多，是否有更简单的方法？

**Comment by Yorkking** [comment]
短路运算和前面有没有 if 没有关系，你单独考虑短路运算就行。

**Reply by hodanna7** [reply]
意思是说只在二元操作中的&&和||考虑短路就行了吗？

**Reply by Yorkking** [reply]
是的

**Reply by hodanna7** [reply]
呜呜谢谢助教

**Reply by junshen1** [reply]
> 短路运算和前面有没有 if 没有关系，你单独考虑短路运算就行。

师兄，咋判断表达式是否为真呢，或者怎么判断类似Builder.CreateICmp(llvm::ICmpInst::ICMP_SLT, lhs, rhs,"lesstmp");返回结果是什么呢？QAQ

**Reply by Yorkking** [reply]
就是类似 if else 去创建跳转。值的获取可能需要使用到 Phi 指令。

**Reply by junshen1** [reply]
还是不太理解怎么根据这个i1的cond判断真假。。

**Reply by Yorkking** [reply]
创建一个 lhs.true 块，创建一个跳转 condbr 到 lhs.true 块：即 lhs 为 true 的时候跳过去，lhs 为false 跳到后面。然后 lhs.true 这个基本块里面放 rhs 生成的 ir。你试着看一下 clang 生成的 && 运算生成的 ir 。

**Reply by Aurora-1412** [reply]
> > 短路运算和前面有没有 if 没有关系，你单独考虑短路运算就行。
> 
> 师兄，咋判断表达式是否为真呢，或者怎么判断类似Builder.CreateICmp(llvm::ICmpInst::ICMP_SLT, lhs, rhs,"lesstmp");返回结果是什么呢？QAQ

泪目，我的疑问和这层楼里的疑问一模一样，思考了一天了感觉不能这么做

**Reply by Yorkking** [reply]
以 A && B 举例：
- 假设当前所在块是 oldBB, 在当前块里： auto value1= build(A) ， 然后创建基本块 land.lhs.true 和 merge 块；
- 根据设置条件跳转 value 为 true 时跳到 land.lhs.true 块，否则跳到后面去 merge;
- 然后在 land.lhs.true 里面生成 B 的代码，比如  auto value2 = build(B)，最后无条件跳转到 merge;
- 最后是 merge 块的处理，merge 块里面是要生成这个逻辑表达式的值：
     - 如果从 oldBB 跳过来就取 value1 的值
     - 如果从 land.lhs.true 跳过来就取 value2 的值
上述的操作，LLVM/IR 里面恰好由 Phi 指令可以完成。

参考：https://releases.llvm.org/11.0.1/docs/tutorial/MyFirstLanguageFrontend/LangImpl05.html#id8

**Reply by hodanna7** [reply]
师兄，我之前用了另一种方法实现，没有用phi，但是效率很低，现在我想重新写一下，不过好像没找到phi指令应该是用哪个API生成的？

**Reply by Yorkking** [reply]
> 师兄，我之前用了另一种方法实现，没有用phi，但是效率很低，现在我想重新写一下，不过好像没找到phi指令应该是用哪个API生成的？

Builder.CreatePHI()

**Comment by Aurora-1412** [comment]
CreatePhi
Phi-&gt;AddInComing
可以查查这两个api(？



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2022年5月26日(周四) 中午11:31
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [arcsysu/SYsU-lang] 短路如何处理 (Discussion #49)




 
师兄，我之前用了另一种方法实现，没有用phi，但是效率很低，现在我想重新写一下，不过好像没找到phi指令应该是用哪个API生成的？
